constraint/system.go

```diff
@@ -0,0 +1,322 @@
+package constraint
+
+import (
+	"fmt"
+	"io"
+	"math/big"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"github.com/consensys/gnark"
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend"
+	"github.com/consensys/gnark/backend/hint"
+	"github.com/consensys/gnark/backend/witness"
+	"github.com/consensys/gnark/debug"
+	"github.com/consensys/gnark/internal/tinyfield"
+	"github.com/consensys/gnark/internal/utils"
+	"github.com/consensys/gnark/logger"
+)
+
+// ConstraintSystem interface that all constraint systems implement.
+type ConstraintSystem interface {
+	io.WriterTo
+	io.ReaderFrom
+	CoeffEngine
+
+	// IsSolved returns nil if given witness solves the constraint system and error otherwise
+	IsSolved(witness *witness.Witness, opts ...backend.ProverOption) error
+
+	// GetNbVariables return number of internal, secret and public Variables
+	// Deprecated: use GetNbSecretVariables() instead
+	GetNbVariables() (internal, secret, public int)
+
+	GetNbInternalVariables() int
+	GetNbSecretVariables() int
+	GetNbPublicVariables() int
+
+	GetNbConstraints() int
+	GetNbCoefficients() int
+
+	Field() *big.Int
+	FieldBitLen() int
+
+	// TODO @gbotrel this should probably go away. check playground usage.
+	// GetSchema() *schema.Schema
+
+	// GetConstraints return a human readable representation of the constraints
+	// TODO @gbotrel restore -- playground uses it.
+	// GetConstraints() [][]string
+
+	AddPublicVariable(name string) int
+	AddSecretVariable(name string) int
+	AddInternalVariable() int
+
+	// AddSolverHint adds a hint to the solver such that the output variables will be computed
+	// using a call to output := f(input...) at solve time.
+	AddSolverHint(f hint.Function, input []LinearExpression, nbOutput int) (internalVariables []int, err error)
+
+	AddCommitment(c Commitment) error
+
+	AddLog(l LogEntry)
+
+	// MakeTerm returns a new Term. The constraint system may store coefficients in a map, so
+	// calls to this function will grow the memory usage of the constraint system.
+	MakeTerm(coeff *Coeff, variableID int) Term
+
+	// AttachDebugInfo enables attaching debug information to multiple constraints.
+	// This is more efficient than using the AddConstraint(.., debugInfo) since it will store the
+	// debug information only once.
+	AttachDebugInfo(debugInfo DebugInfo, constraintID []int)
+
+	IsValid() error // TODO @gbotrel should take list of Validators
+}
+
+// CoeffEngine capability to perform arithmetic on Coeff
+type CoeffEngine interface {
+	FromInterface(interface{}) Coeff
+	ToBigInt(*Coeff) *big.Int
+	Mul(a, b *Coeff)
+	Add(a, b *Coeff)
+	Sub(a, b *Coeff)
+	Neg(a *Coeff)
+	Inverse(a *Coeff)
+	One() Coeff
+	IsOne(*Coeff) bool
+	String(*Coeff) string
+}
+
+type Iterable interface {
+	// WireIterator returns a new iterator to iterate over the wires of the implementer (usually, a constraint)
+	WireIterator() func() int
```

Comment:
ha, channel would be pretty but I think it's going to be memory innefficient. Since we operate with one thread only so far in the frontend, we need to "fill" the channel with all the wireID values. So iterating over large linear expression, at each constraint, will force us to allocate the memory for the channel. 

The current pattern is a "traditional" `next()` function which returns -1 when it's done iterating over. Will document.

Reply:
Or func() (int, bool), where bool indicates if there is another element coming?